cpp/src/gandiva/llvm_generator.cc

```diff
@@ -953,25 +1021,45 @@ LValuePtr LLVMGenerator::Visitor::BuildValueAndValidity(const ValueValidityPair&
 }
 
 LValuePtr LLVMGenerator::Visitor::BuildFunctionCall(const NativeFunction* func,
+                                                    DataTypePtr arrow_return_type,
                                                     std::vector<llvm::Value*>* params) {
-  auto arrow_return_type = func->signature().ret_type()->id();
-  auto llvm_return_type = generator_->types()->IRType(arrow_return_type);
-
-  // add extra arg for return length for variable len return types (alloced on stack).
-  llvm::AllocaInst* result_len_ptr = nullptr;
-  if (arrow::is_binary_like(arrow_return_type)) {
-    result_len_ptr = new llvm::AllocaInst(generator_->types()->i32_type(), 0,
-                                          "result_len", entry_block_);
-    params->push_back(result_len_ptr);
-    has_arena_allocs_ = true;
-  }
+  auto types = generator_->types();
+  auto arrow_return_type_id = arrow_return_type->id();
+  auto llvm_return_type = types->IRType(arrow_return_type_id);
+
+  if (arrow_return_type_id == arrow::Type::DECIMAL) {
+    // For decimals, the output struct pointer is passed along as an input
+    // param to the function call. This allows for a stack allocation instead
+    // of a malloc/new.
```

Comment:
I'm not sure what you mean. AFAIK you should be able to return a struct type from a LLVM function.

Reply:
but, that would be an extra copy. And, I still need to pass along the result precision/scale to the LLVM function (I don't want it to recompute for each record).

Reply:
Ok. The choice is between 

(a) struct_decimal result = add(struct_decimal x, struct_decimal y, int32_t result_precision, int32_t result_scale)

vs

(b) void add(const struct_decimal *x, const struct_decimal *y, struct_decimal *result)

I like b better :-).

Reply:
do let me know if you prefer (a) strongly - I'll switch to (a).

Reply:
No, it's fine.